PSMDB-1372 Pro builds documentation (6.0) (#788)
percona/psmdb-docs: docs/install/source.md

# Build from source code

This document guides you though the steps how to build Percona Server for MongoDB from source code.

## Available builds

- Pro builds. These builds include solutions such as FIPS mode that are typically demanded by large enterprises. They are included into packages provided by Percona and are available to Percona Customers. [Learn how to become a Customer](https://www.percona.com/about/contact). 

> - Basic builds. These include all Percona Server for MongoDB functionality except the solutions in pro builds. The packages provided by Percona are available to anyone.

Reviewer: shouldn't we capitalize Pro builds as the name?

### Prepare the build environment

> 1. Create the folder where all build actions take place. For the steps below we use the `/tmp/Percona Server for MongoDB/test` folder.

Reviewer 2: ```suggestion
1. Create the folder where all build actions take place. For the steps below we use the `/tmp/psmdb/test` folder.
```

2. Navigate to the build folder and download the build script. Replace the `<tag>` placeholder with the required version of Percona Server for MongoDB:

    ```{.bash data-prompt="$"}

>     $ wget https://raw.githubusercontent.com/percona/percona-server-mongodb/Percona Server for MongoDB-<tag>/percona-packaging/scripts/Percona_Server_for_MongoDB_builder.sh -O Percona_Server_for_MongoDB_builder.sh 

Reviewer 2: ```suggestion
    $ wget https://raw.githubusercontent.com/percona/percona-server-mongodb/psmdb-<tag>/percona-packaging/scripts/psmdb_builder.sh -O psmdb_builder.sh 
```

    === "Basic build"    

        ```{.bash data-prompt="$"}
        $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb rhel:9 sh -c '
        set -o xtrace
        cd /tmp/psmdb
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
        --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1
        '
        ```    

    === "Pro build"    

        ```{.bash data-prompt="$"}
        $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb rhel:9 sh -c '
        set -o xtrace
        cd /tmp/psmdb
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \

>         --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --enable_fipsmode=1 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1

Reviewer 2: ```suggestion
        --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1
```

### Build steps

Use the following instructions to build [tarballs](#tarballs) or [packages](#packages):

#### Tarballs

1. The following command builds tarballs of Percona Server for MongoDB 6.0.6-5 on Red Hat Enterprise Linux 9. Change the Docker image and the values for `--branch`, `--psm_ver`, `--psm_release` flags to build tarballs of a different version and on a different operating system.

    === "Basic build"    

        ```{.bash data-prompt="$"}

>         $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb rhel:9 sh -c '

Reviewer 2: ```suggestion
        $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '
```

        set -o xtrace
        cd /tmp/psmdb
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
        --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1

>         '

Reviewer 2: ```suggestion
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
        --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1 --build_tarball=1
        '
```

        ```    

>     === "Pro build"    

Reviewer 2: ```suggestion
```

We don't provide tarballs for PRO builds

>         ```{.bash data-prompt="$"}
>         $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb rhel:9 sh -c '
>         set -o xtrace
>         cd /tmp/psmdb
>         bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
>         bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
>         --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1

Reviewer 2: ```suggestion
```

        '
        ```    

    The command does the following:    

    * runs Docker daemon as the root user using the RHEL 9 image
    * mounts the build directory into the container
    * establishes the shell session inside the container
    * inside the container, navigates to the build directory and runs the build script to install dependencies 
    * runs the build script again to build the tarball for the Percona Server for MongoDB version 6.0.6-5

2. Check that tarballs are built:

    ```{.bash data-prompt="$"}

>     $ ls -la /tmp/psmdb/test/source_tarball/

Reviewer 2: ```suggestion
    $ ls -la /tmp/psmdb/test/tarball/
```

### Build steps

Use the following instructions to build [tarballs](#tarballs) or [packages](#packages):

#### Tarballs

!!! note

    You can build only Percona Server for MongoDB basic tarballs with the build script. Percona Server for MongoDB Pro tarballs are not supported.

> 1. The following command builds tarballs of Percona Server for MongoDB 6.0.12-9 on Oracle Linux 7. Change the Docker image and the values for `--branch`, `--psm_ver`, `--psm_release` flags to build tarballs of a different version and on a different operating system.

Reviewer 2: ```suggestion
1. The following command builds tarballs of Percona Server for MongoDB 6.0.12-9 on Centos 7. Change the Docker image and the values for `--branch`, `--psm_ver`, `--psm_release` flags to build tarballs of a different version and on a different operating system.
```

    ```{.bash data-prompt="$"}

>     $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '

Reviewer 2: ```suggestion
    $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb centos:7 sh -c '
```

    set -o xtrace
    cd /tmp/psmdb
    bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
    bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
    --branch=release-6.0.12-9 --psm_ver=6.0.12 --psm_release=9 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1 --build_tarball=1
    '
    ```    

    The command does the following:    

>     * runs Docker daemon as the root user using the Oracle Linux 7 image

Reviewer 2: ```suggestion
    * runs Docker daemon as the root user using the Centos 7 image
```

    * mounts the build directory into the container
    * establishes the shell session inside the container
    * inside the container, navigates to the build directory and runs the build script to install dependencies 
    * runs the build script again to build the tarball for the Percona Server for MongoDB version 6.0.12-9

2. Check that tarballs are built:

    ```{.bash data-prompt="$"}
    $ ls -la /tmp/psmdb/test/tarball/
    ```

    Sample output:

    ```{.text .no-copy}
    total 88292
    -rw-r--r--. 1 root root 90398894 Jul  1 10:58 percona-server-mongodb-6.0.12-9-x86_64.glibc2.17.tar.gz
    ```

#### Packages

> 1. Build the source tarball. It serves as the base for source packages. It is important to build source tarball using the oldest supported operating system, which is Oracle Linux 7.

Reviewer 2: ```suggestion
1. Build the source tarball. It serves as the base for source packages. It is important to build source tarball using the oldest supported operating system, which is Centos 7.
```

    ```{.bash data-prompt="$"}

>     $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '

Reviewer 2: ```suggestion
    $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb centos:7 sh -c '
```

    set -o xtrace
    cd /tmp/psmdb
    bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
    bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git --branch=release-6.0.12-9 --psm_ver=6.0.12 --psm_release=9 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1'
    ```

2. Build source packages. These packages include the source code and patches and are used to build binary packages.

>     Note that to build source packages you still have to use the oldest supported operating system: Oracle Linux 7 for RPMs and Ubuntu 18.04 (Bionic Beaver) for DEB packages. 

Reviewer 2: ```suggestion
    Note that to build source packages you still have to use the oldest supported operating system: Centos 7 for RPMs and Ubuntu 18.04 (Bionic Beaver) for DEB packages. 
```

    === "Basic build" 

        === "DEB"    

            ```{.bash data-prompt="$"}
            $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb ubuntu:bionic sh -c '
            set -o xtrace
            cd /tmp/psmdb
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
            --branch=release-6.0.12-9 --psm_ver=6.0.12--psm_release=9 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --build_src_deb=1
            '
            ```    

            Check that source packages are created    

            ```{.bash data-prompt="$"}
            $ ls -la /tmp/psmdb/test/source_deb/
            ```    

            Sample output:    

            ```{.text .no-copy}
            rw-r--r--. 1 root root 90398894 Jul  1 11:45 percona-server-mongodb_6.0.12.orig.tar.gz
            ```
            
        === "RPM"    

            ```{.bash data-prompt="$"}

>             $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '

Reviewer 2: ```suggestion
            $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb centos:7 sh -c '
```

            rw-r--r--. 1 root root 90398894 Jul  1 11:45 percona-server-mongodb-6.0.12-9.generic.src.rpm
            ```

    === "Pro build" 

        === "DEB"    

            ```{.bash data-prompt="$"}
            $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb ubuntu:bionic sh -c '
            set -o xtrace
            cd /tmp/psmdb
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
            --branch=release-6.0.12-9 --psm_ver=6.0.12--psm_release=9 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_src_deb=1
            '
            ```    

            Check that source packages are created    

            ```{.bash data-prompt="$"}
            $ ls -la /tmp/psmdb/test/source_deb/
            ```    

            Sample output:    

            ```{.text .no-copy}
            rw-r--r--. 1 root root 90398894 Jul  1 11:45 percona-server-mongodb-pro_6.0.12.orig.tar.gz
            ```
            
        === "RPM"    

            ```{.bash data-prompt="$"}

>             $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '

Reviewer 2: ```suggestion
            $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb centos:7 sh -c '
```

>     -rw-r--r--. 1 root root 90398894 Jul  1 10:58 percona-server-mongodb-6.0.6-5.tar.gz

Reviewer 2: ```suggestion
    -rw-r--r--. 1 root root 90398894 Jul  1 10:58 percona-server-mongodb-6.0.12-9-x86_64.glibc2.17.tar.gz
```

    ```

#### Packages

> 1. Build source packages

Reviewer 2: ```suggestion
1. Build src tarball
docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '
set -o xtrace
cd /tmp/psmdb
bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --get_sources=1'

2. Build source packages
```

    === "DEB"

        ```{.bash data-prompt="$"}
        $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb ubuntu:jammy sh -c '
        set -o xtrace
        cd /tmp/psmdb
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
        bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
        --branch=release-6.0.6-5 --psm_ver=6.0.6--psm_release=5 --mongo_tools_tag=100.7.0 --jemalloc_tag=psmdb-3.2.11-3.1 --build_src_deb=1
        '
        ```

        Check that source packages are created

        ```{.bash data-prompt="$"}
        $ ls -la /tmp/psmdb/test/source_deb/
        ```

        Sample output:

        ```{.text .no-copy}
        rw-r--r--. 1 root root 90398894 Jul  1 11:45 percona-server-mongodb_6.0.6.orig.tar.gz
        ```
        
    === "RPM"

        ```{.bash data-prompt="$"}

>         $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb ubuntu:jammy sh -c '

Reviewer 2: ```suggestion
        $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:7 sh -c '
```

            cd /tmp/psmdb
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \

>             --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --enable_fipsmode=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_deb=1

Reviewer 2: ```suggestion
            --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_deb=1
```

            cd /tmp/psmdb
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \

>             --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --enable_fipsmode=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_rpm=1

Reviewer 2: ```suggestion
            --branch=release-6.0.6-5 --psm_ver=6.0.6 --psm_release=5 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_rpm=1
```

            Sample output:

            ```{.text .no-copy}

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-dbg-pro_6.0.12-9.jammy_amd64.deb  

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-pro-dbg_6.0.12-9.jammy_amd64.deb  
```

            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-mongos-pro_6.0.12-9.jammy_amd64.deb 
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-server-pro_6.0.12-9.jammy_amd64.deb 

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-pro_6.0.12-9.jammy_amd64.deb  

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-6.0.12-9.jammy_amd64.deb  
```

            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-pro_6.0.12-9.jammy_amd64.deb
            ```
        
        === "RPM"

            ```{.bash data-prompt="$"}
            $ docker run -ti -u root -v /tmp/psmdb:/tmp/psmdb oraclelinux:9 sh -c '
            set -o xtrace
            cd /tmp/psmdb
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --install_deps=1
            bash -x ./psmdb_builder.sh --builddir=/tmp/psmdb/test --repo=https://github.com/percona/percona-server-mongodb.git \
            --branch=release-6.0.12-9 --psm_ver=6.0.12 --psm_release=9 --mongo_tools_tag=100.7.0 --full-featured=1 --jemalloc_tag=psmdb-3.2.11-3.1 --build_rpm=1
            '
            ```

            Check that source packages are created

            ```{.bash data-prompt="$"}
            $ ls -la /tmp/psmdb/test/rpm/
            ```

            Sample output:

            ```{.text .no-copy}
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-pro-6.0.12-9.el8.x86_64.rpm  

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-debugsource-pro-6.0.12-9.el8.x86_64.rpm 

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-pro-debugsource-6.0.12-9.el8.x86_64.rpm 
```

            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-mongos-pro-6.0.12-9.el8.x86_64.rpm    

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-mongos-debuginfo-pro-6.0.12-9.el8.x86_64.rpm 

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-mongos-pro-debuginfo-6.0.12-9.el8.x86_64.rpm 
```

            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-server-pro-6.0.12-9.el8.x86_64.rpm    

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-server-debuginfo-pro-6.0.12-9.el8.x86_64.rpm 

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-server-pro-debuginfo-6.0.12-9.el8.x86_64.rpm 
```

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-pro-6.0.12-9.el8.x86_64.rpm 

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-6.0.12-9.el8.x86_64.rpm 
```

>             rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-debuginfo-pro-6.0.12-9.el8.x86_64.rpm

Reviewer 2: ```suggestion
            rw-r--r--. 1 root root 90398894 Jul  1 13:16 percona-server-mongodb-tools-debuginfo-6.0.12-9.el8.x86_64.rpm
```
